File: Linux/get-info.py
#!/usr/bin/env python3

import json
import sys
import re
import os
import tempfile
from subprocess import PIPE, DEVNULL, Popen
from typing import Optional, List
from config import arcconf_path

"""
Description: Get Information From Adaptec RAID Controller.
Required external program: arcconf

Version: 1
Author: Chizhov Nikolay
E-Mail: nchizhov@inok.ru
"""

title_split = '----------------------------------------------------------------------'
subtitle_split = '--------------------------------------------------------'
logical_device_re = re.compile('^Logical device number (\d+)$')
physical_device_re = re.compile('^Device #(\d+)$')


def adapter_data(index: int) -> Optional[List[str]]:
    arcconf_process = Popen('{path} GETCONFIG {index} AL'.format(path=arcconf_path, index=index),
                            shell=True, stdout=PIPE, stderr=DEVNULL)
    arcconf_lines = arcconf_process.communicate()[0].decode('ascii').splitlines()
    if arcconf_process.returncode != 0:
        return None
    return arcconf_lines


def parse_data(data: List[str]) -> dict:
    controller_data = {}
    current_section = None
    current_subsection = None
    current_index = None
    for idx, line in enumerate(data):
        line = line.strip()
        if idx < 3:
            continue
        if line == title_split and data[idx - 2].strip() == title_split:
            current_section = data[idx - 1].strip()
            current_subsection = None
            current_index = None
            controller_data[current_section] = {}
            continue
        if line == subtitle_split and data[idx - 2].strip() == subtitle_split:
            current_subsection = data[idx - 1].strip()
            if current_index is None:
                controller_data[current_section][current_subsection] = {}
            else:
                controller_data[current_section][current_index][current_subsection] = {}
            continue
        logical_device = logical_device_re.search(line)
        if logical_device is not None:
            current_index = logical_device[1]
            current_subsection = None
            controller_data[current_section][current_index] = {}
            continue
        physical_device = physical_device_re.search(line)
        if physical_device is not None:
            current_index = physical_device[1]
            current_subsection = None
            controller_data[current_section][current_index] = {
                'Device Info': data[idx + 1].strip()
            }
            continue
        s_line = line.split(': ', 1)
        if len(s_line) == 2:
            field = s_line[0].strip()
            field_val = s_line[1].strip()
            if current_index is None:
                if current_subsection is None:
                    controller_data[current_section][field] = field_val
                else:
                    controller_data[current_section][current_subsection][field] = field_val
            else:
                if current_subsection is None:
                    controller_data[current_section][current_index][field] = field_val
                else:
                    controller_data[current_section][current_index][current_subsection][field] = field_val
    return controller_data


def save_info(index: int, data: dict) -> None:
    info_path = os.path.join(tempfile.gettempdir(), 'adaptec-{index}.json'.format(index=index))
    if os.path.exists(info_path):
        os.remove(info_path)
    with open(info_path, 'w') as f:
        json.dump(data, f, ensure_ascii=False)


if __name__ == '__main__':
    if len(sys.argv) != 2 or not sys.argv[1].isnumeric():
        sys.exit(1)
    index = int(sys.argv[1])
    data = adapter_data(index)
    if data is None:
        sys.exit(1)
    parsed_data = parse_data(data)
    save_info(index, parsed_data)

File: Util/adapter/get_info.go
package adapter

import (
	"encoding/json"
	"fmt"
	"log"
	"os"
	"path"
	"regexp"
	"strconv"
	"strings"
	"support"
)

func adapterData(configUtil string, index uint) (string, error) {
	return support.RunCommand(configUtil, "GETCONFIG", strconv.FormatUint(uint64(index), 10), "AL")
}

func parseData(data []string) map[string]any {
	titleSplit := "----------------------------------------------------------------------"
	subtitleSplit := "--------------------------------------------------------"
	logicalDeviceRe, _ := regexp.Compile(`^Logical device number (\d+)$`)
	physicalDeviceRe, _ := regexp.Compile(`^Device #(\d+)$`)

	controllerData := make(map[string]any)
	currentSection := support.NewNil()
	currentSubsection := support.NewNil()
	currentIndex := support.NewNil()
	for idx, line := range data {
		if idx < 3 {
			continue
		}
		line = strings.TrimSpace(line)
		if line == titleSplit && strings.TrimSpace(data[idx-2]) == titleSplit {
			currentSection = support.NewData(strings.TrimSpace(data[idx-1]))
			currentSubsection = support.NewNil()
			currentIndex = support.NewNil()
			support.AddDeepNested(controllerData, []string{currentSection.Value().(string)}, make(map[string]any))
			continue
		}
		if line == subtitleSplit && strings.TrimSpace(data[idx-2]) == subtitleSplit {
			currentSubsection = support.NewData(strings.TrimSpace(data[idx-1]))
			sectionValue := currentSection.Value().(string)
			subsectionValue := currentSubsection.Value().(string)
			if currentIndex.Value() == nil {
				support.AddDeepNested(controllerData,
					[]string{sectionValue, subsectionValue}, make(map[string]any))
			} else {
				support.AddDeepNested(controllerData,
					[]string{sectionValue, currentIndex.Value().(string), subsectionValue}, make(map[string]any))
			}
			continue
		}
		sectionValue := currentSection.Value().(string)
		logicalDevice := logicalDeviceRe.FindSubmatch([]byte(line))
		if logicalDevice != nil {
			indexValue := string(logicalDevice[1])
			currentIndex = support.NewData(indexValue)
			currentSubsection = support.NewNil()
			support.AddDeepNested(controllerData, []string{sectionValue, indexValue}, make(map[string]any))
			continue
		}
		physicalDevice := physicalDeviceRe.FindSubmatch([]byte(line))
		if physicalDevice != nil {
			indexValue := string(physicalDevice[1])
			currentIndex = support.NewData(indexValue)
			currentSubsection = support.NewNil()
			support.AddDeepNested(controllerData,
				[]string{sectionValue, indexValue, "Device Info"}, strings.TrimSpace(data[idx+1]))
			continue
		}
		sLine := strings.SplitN(line, ": ", 2)
		if len(sLine) != 2 {
			continue
		}
		field := strings.TrimSpace(sLine[0])
		fieldVal := strings.TrimSpace(sLine[1])
		if currentIndex.Value() == nil {
			if currentSubsection.Value() == nil {
				support.AddDeepNested(controllerData,
					[]string{sectionValue, field}, fieldVal)
			} else {
				support.AddDeepNested(controllerData,
					[]string{sectionValue, currentSubsection.Value().(string), field}, fieldVal)
			}
			continue
		}
		indexValue := currentIndex.Value().(string)
		if currentSubsection.Value() == nil {
			support.AddDeepNested(controllerData,
				[]string{sectionValue, indexValue, field}, fieldVal)
		} else {
			support.AddDeepNested(controllerData,
				[]string{sectionValue, indexValue, currentSubsection.Value().(string), field}, fieldVal)
		}
	}
	return controllerData
}

func saveInfo(index uint, data map[string]any) {
	infoPath := path.Join(os.TempDir(), fmt.Sprintf("adaptec-%d.json", index))
	if support.FileExists(infoPath) {
		err := os.Remove(infoPath)
		if err != nil {
			return
		}
	}
	jsonStr, err := json.Marshal(data)
	if err != nil {
		return
	}
	err1 := os.WriteFile(infoPath, jsonStr, 0644)
	if err1 != nil {
		return
	}
}

func GetInfo(args support.ArgsList) {
	config, err := support.ReadConfig()
	if err != nil {
		log.Fatalln(err)
	}

	aData, err := adapterData(config.ArcconfPath, args.Index)
	if err != nil {
		log.Fatalln(err)
	}

	data := strings.Split(aData, "\n")
	controllerData := parseData(data)

	saveInfo(args.Index, controllerData)
	os.Exit(0)
}
